front: import train sets in scenario from catalog

sim51 opened the pull request:
[Link to design](https://www.sketch.com/s/008d6980-d719-4bdc-bb2a-b4c2dda0dc1e/v/ln1K7z/p/5BF717FB-5D63-4A0E-8079-5E4E69C235E0/)

[Link to issue](https://github.com/osrd-project/osrd-confidential/issues/758)

All data are mocked in the dialog, and the submit action is not done. You can check this [TODO](https://github.com/OpenRailAssociation/osrd/blob/bsi/tg-catalog/front/src/applications/operationalStudies/views/Scenario/components/ManageTimetableItem/ManageTimetableItemModal.tsx#L68)
But the design and user's interactions are done

sim51 committed:
front: import train sets in scenario from catalog

For now, data are mocked, but design and interaction are done.

Signed-off-by: Benoit Simard <contact@bsimard.com>

Yohh commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/useLoadCatalog.ts:
```diff
@@ -0,0 +1,71 @@
+      // Build the mapping of catalog entries with train schedule sets
+      const catalogEntries: Map<CatalogEntry['id'], Array<TrainScheduleSet['id']>> = new Map();
+      trainScheduleSets
+        .filter((e) => e.catalog_entry_id && e.published)
```
nit
```suggestion
        .filter((set) => set.catalog_entry_id && set.published)
```

Yohh commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/useLoadCatalog.ts:
```diff
@@ -0,0 +1,71 @@
+import { useCallback, useEffect, useState } from 'react';
+
+import { type CatalogEntry, type TrainScheduleSet } from 'common/api/osrdEditoastApi';
```
nit
```suggestion
import type { CatalogEntry, TrainScheduleSet } from 'common/api/osrdEditoastApi';
```

SarahBellaha commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,111 @@
+      case 'copy':
+        return <NoEntry variant="fill" className={className} />;
+      case 'reference':
+        return <CheckCircle variant="fill" className={className} />;
```
```suggestion
        return <CheckCircle variant="fill" className={className} iconColor="var(--info30)"  />;
```
This icon should be blue :)

Yohh commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/useLoadCatalog.ts:
```diff
@@ -0,0 +1,71 @@
+          if (!catalog_entry)
+            throw new Error(
+              `Can't find catalog ${catalog_id} but used in the following train schedule sets : ${trainScheduleSetIds.join(',')} `
+            );
```
even if I set `catalog_entry_id` to `null` in `mockTrainSchedulesSets.ts` I don't get this error, nothing happens

also I wonder why `catalog_entry_id` can be null...

sim51 replied:
A user can also create a `TrainScheduleSet` locally, and in this case the  `catalog_entry_id` is null  (it's not a reference). 

About the error, to have it we need to have a `TrainScheduleSet` with a `catalog_entry_id` that doesn't exist.
It should never happened, except if we have an integrity error, or if a catalog entry has been deleted while loading the train schedule sets.

I can change the line 36 by `const catalog_entry = await mockGetCatalogEntry(catalog_id)!;`  because I'm 99.9..% sure to obtain a `CatalogEntry` but I prefer to handle the case with an error.

SarahBellaha commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetCatalogDialog.tsx:
Should we rename this file `TrainScheduleSetCatalogDialog.tsx` ?

sim51 replied:
In this case, I prefer to : 
- Rename the `index.tsx` to `TrainScheduleSetCatalogDialog.tx` 
- Create an `index.tsx` that export the `TrainScheduleSetCatalogDialog` (and not as default)

With such a structure, we clearly know that the main component is a `Dialog` and it's easy to navigate to it.

What do you think ?

sim51 replied:
Done

SarahBellaha reviewed (commented):
<img width="1551" height="246" alt="image" src="https://github.com/user-attachments/assets/36334d76-1e63-4bb5-b2f2-33583e0fff14" />

The `trainScheduleSet.name` misses the "Package / paquet" prefix in the right list :)

Quick question @thibautsailly : 
When should a package (reference) be checked in this view ? My guess is that it indicates that the package reference we chose is already linked to our scenario, is that right ?

Yohh commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,111 @@
+          getOptionLabel={(e) => tos(`importType.${e}`)}
+          getOptionValue={(e) => `${e}`}
+          getOptionIcon={(e) => (e === 'copy' ? <DesktopDownload /> : <Broadcast />)}
+          onChange={(e) => addToCart(trainScheduleSet.id, e)}
```
```suggestion
          value={importType}
          getOptionLabel={(type) => tos(`importType.${type}`)}
          getOptionValue={(type) => `${type}`}
          getOptionIcon={(type) => (type === 'copy' ? <DesktopDownload /> : <Broadcast />)}
          onChange={(type) => addToCart(trainScheduleSet.id, type)}
```

Yohh commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetsCart.tsx:
```diff
@@ -0,0 +1,55 @@
+  const { t } = useTranslation('operational-studies', { keyPrefix: 'importTrainscheduleSet' });
+
+  /**
+   * List of TrainSchedule set in the cart
```
nit
```suggestion
   * List of TrainSchedule sets in the cart
```

Yohh commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetsCatalog.tsx:
```diff
@@ -0,0 +1,99 @@
+};
+
+const Collapsable = (props: CollapsableProps) => {
+  const { children, ...divProps } = props;
```
why creating an isolated `Collapsable` component?

this way we have to manage children and divProps, it seems to me less readable, this component isn't exported so I don't understand the purpose

is it to anticipate something?

SarahBellaha replied:
I think we could use the `<Collapsable />` component from the common folder as they appear to do the same thing.
You can pass the new icons in the `iconOpen` and `iconClose` props and add some classNames to adjust the styling :)

sim51 replied:
At the beginning I was hopping to reuse the component `src/common/Collapsable`, but due to the design, I needed to redo it. 

IMHO, it's more readable to have many small components than a big one, even if they are used at one place (ie. not exported).

Yohh replied:
it's not about the fact that we have multiple components, it's about the props and children management, it seems to me too much complicated in this case than a single component, it may be just a matter of preference

sim51 replied:
I did a commit to reuse the common `Collapse` component. 
I changed it and did some modifications on the `IncompatibleConstraints` components  (which were impacted by the modif).

Yohh commented on front/public/locales/en/operational-studies.json:
```diff
@@ -54,6 +54,16 @@
+    "import": "Import",
+    "importType": {
+      "copy": "Import by copy",
+      "reference": "Import by refrence"
```
```suggestion
      "reference": "Import by reference"
```

Yohh commented:
the import button miss some style:

<img width="154" height="157" alt="image" src="https://github.com/user-attachments/assets/287d964e-985b-49d2-97e5-b67ed108796e" />
VS
<img width="171" height="174" alt="image" src="https://github.com/user-attachments/assets/9635a7dd-c426-4363-ab00-badfd90d3d7a" />
<hr/>

<img width="251" height="172" alt="image" src="https://github.com/user-attachments/assets/7aa78ffe-1cb5-4f29-b55d-47911d210bb8" />
VS
<img width="382" height="290" alt="image" src="https://github.com/user-attachments/assets/b73f2fb6-46f9-4a3d-a4d2-87e327457fc0" />

Yohh commented:
modal title is different in the mockup

<img width="253" height="58" alt="image" src="https://github.com/user-attachments/assets/523578a5-ec2f-4463-9f5a-acba91d6db45" />
VS
<img width="519" height="88" alt="image" src="https://github.com/user-attachments/assets/d939c979-ab47-4c35-8ff6-25a55d1f53e3" />

Yohh commented:
<img width="187" height="29" alt="image" src="https://github.com/user-attachments/assets/a1707f58-a1a7-41cb-ad8d-15ebfb1a9447" />

needs `semi-bold-600`

Yohh commented:
<img width="208" height="29" alt="image" src="https://github.com/user-attachments/assets/9af9fc75-90eb-451c-903b-89f40a558147" />

the right icon needs to be vertically aligned

<img width="250" height="49" alt="image" src="https://github.com/user-attachments/assets/1a5c98ed-3bce-4338-b42a-ad1459347481" />

sim51 commented:
I fixed all your comments except the one for the button. 
I need to check with @thibautsailly if the others buttons should have the same design, or if this button is special.
Will also check with him when I should display the reference in full blue/green.

sim51 commented:
About the button, @thibautsailly  has updated the design on sketch.
For the blue/green color, yes it is present when the `TimeScheduleSet` is already part of the scenario.
Will do the changes tomorrow

sim51 commented:
For the blue/green color, it is used for TSS that are already imported on the timetable, and we can only know it for reference.  Due to that, the element is not selectable (I added a title to explain it on the disable button).
I added a new props to the dialog for that (to know which TSS to disabled).

I also did the refacto for the `index.tsx`, so we should be good now.

SharglutDev commented on front/public/locales/en/operational-studies.json:
```diff
@@ -54,6 +54,18 @@
+      "copy": "Import by copy",
+      "reference": "Import by reference"
+    },
+    "package": "Package",
```
Same for fr

```suggestion
    "trainScheduleSet": "Package",
```

sim51 replied:
Done

SharglutDev commented on front/public/locales/en/operational-studies.json:
```diff
@@ -54,6 +54,18 @@
+      "reference": "Import by reference"
+    },
+    "package": "Package",
+    "title": "Import train group versions",
```
I think we don't want to use the "group" term.
We could use something like "package sets". Is it ok for you @maelysLeratRosso ?

```suggestion
    "title": "Import train package sets",
```

maelysLeratRosso replied:
I would say "Import packages" directly (or possibly "Import train packages") but I think package and sets are redundant

sim51 replied:
Done for "Import train packages"

SharglutDev commented on front/public/locales/fr/operational-studies.json:
```diff
@@ -54,6 +54,18 @@
+      "reference": "Import par référence"
+    },
+    "package": "Paquet",
+    "title": "Import d'ensembles de train",
```
```suggestion
    "title": "Import d'ensemble de paquets de trains",
```

maelysLeratRosso replied:
You should write "Import de paquets de trains"

sim51 replied:
Done

SharglutDev commented on front/public/locales/fr/operational-studies.json:
```diff
@@ -217,7 +229,9 @@
-      "importTimetableItem": "Importer des trains",
+      "importTimetableFromCatalog": "Importer depuis le catalogue",
+      "importTimetableFromFile": "Importer depuis un fichier",
+      "importTimetableItem": "Importer une grille horaire",
```
Why change this one ? Was it requested by @maelysLeratRosso or @thibautsailly ?
If yes, we should also update the english one.

maelysLeratRosso replied:
I see it in the mockup
In English it should be "Import from file"

sim51 replied:
@SharglutDev  is about the key `importTimetableItem`.
I don't remember why I changed it, and I can't find trace of this change neither in the feature ticket, neither in sketch.  Will do the change.

About others keys, yes it's in sketch.

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
nit: `TrainScheduleSet` is like a "key" word for us in the app. I'm not sure we should use it to name a file like that.

Maybe we could have 2 files for each component in this one ?

sim51 replied:
I splitted the file in two : 
- TrainScheduleSetCartItem
- TrainScheduleSetMenuItem

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+          </div>
+          <div className="metadata">{trainScheduleSet.published ? 'published' : 'unpublished'}</div>
+        </div>
+      </button>
```
Was it discussed with @thibautsailly that this specific zone is clickable to add the package in the cart ? 
<img width="541" height="134" alt="Image" src="https://github.com/user-attachments/assets/43020ea1-27f7-4b31-90a4-53c418e119e0" />

Seems to me that it should be at least the whole width of the card "header" or the whole card itself (description included)

sim51 replied:
Done

SharglutDev replied:
Much better thanks !

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+              If the item is imported as a ref, we display the icon, 
+              as well if it is disable because it can only be a ref 
+            */}
+            {(cartImportType === 'reference' || disabled) && <Broadcast className="d-flex" />}
```
For the first case `cartImportType === 'reference'`, I don't see any mention of that in the mockup or in the AC. 
Do we want to show the icon if the package is in the cart by reference ? @maelysLeratRosso

maelysLeratRosso replied:
We can leave it as it is for now, we'll see if we get feedback :)

sim51 replied:
It's what I understood from the ticket and with the sketch. 
Following sarah comment, we talk about it in a call with maelys and the summary is in this comment : https://github.com/OpenRailAssociation/osrd/pull/14461#issuecomment-3729760591

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/useLoadCatalog.ts:
```diff
@@ -0,0 +1,71 @@
+      // Get all catalog entries to have there name
+      const catalog = await Promise.all(
+        catalogEntries.entries().map(async ([catalog_id, trainScheduleSetIds]) => {
+          const catalog_entry = await mockGetCatalogEntry(catalog_id);
```
We already discussed about that but are we sure it is better to call x times "getCatalogEntryById" instead of one time "getCatalogEntries" (with x results) ?

Moreover, in the back end implementation, there is no AC to create a `getCatalogEntryById` endpoint and at the moment, it is not possible to create a catalog_entry without a train_schedule_set inside so all catalog_entry we will get by the endpoint will be used in the dialog.

Do you think this should be challenged ?

sim51 replied:
Yes there is an issue here. By checking the API plan, I see :
- GET /catalogue_entry/ =>  get all the entries
- GET /catalogue_entry/{id}/train_schedule_sets?published=true => to get the published train schedule

So we will still have a N+1 issue here.

Will change it

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetCatalogDialog.tsx:
```diff
@@ -0,0 +1,121 @@
+   * => notify
+   */
+  useEffect(() => {
+    if (error) dispatch(setFailure(castErrorToFailure(error)));
```
After the recent discussions we had about erros in dialog components, should we rather display the error in this dialog footer ?

sim51 replied:
agree!

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetCatalogDialog.tsx:
```diff
@@ -0,0 +1,121 @@
+  /**
+   * Add a TSS to cart.
+   */
+  const addToCart: CartManagement['addToCart'] = useCallback(
```
Maybe we should rename this function to `upsertToCart` because we use it to add a new tss in the cart of to update the tss when switching its type.

sim51 replied:
Done

SharglutDev reviewed (commented):
Thanks for the PR ! 
Here is a first batch of reviews, not finished yet :)

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetsCatalog.tsx:
```diff
@@ -0,0 +1,77 @@
+   * List of catalog entries
+   */
+  const catalogEntries = useMemo(() => {
+    const catalogData = (Object.keys(catalog) as unknown as CatalogEntry['id'][]).map((catId) => {
```
We try to avoid using `as unknown as` in the codebase.
If `CatalogById` was a `Map` rather than a `Record`, we could do `catalog.keys()` and then typescript should infer the typing of those keys.
Moreover, we can access the values with a "xx.get()" and typescript also helps us by typing the value to be possibly `undefined`.
What do you think ?
We can do the same for `TrainScheduleSetById`

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/types.ts:
```diff
@@ -0,0 +1,18 @@
+export type CatalogById = Record<
+  CatalogEntry['id'],
+  CatalogEntry & {
+    trainScheduleSetIds: Array<TrainScheduleSet['id']>;
```
nit: I feel like we do not use a lot this syntax in the codebase
```suggestion
    trainScheduleSetIds: TrainScheduleSet['id'][];
```

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetsCatalog.tsx:
```diff
@@ -0,0 +1,77 @@
+          className="catalog-entry"
+          iconPosition="left"
+          iconOpen={
+            <TriangleDown className="d-flex align-items-center justify-content-center mr-1" />
```
Recently a PR was merged to remove some sncf bootstrap components and mpre globally the sncf bootstrap lib.

We probably want, at some point when all the old UI components won't be used anymore, to totally remove bootstrap from the dependencies (can be discussed in a front meeting) so we should probably avoid introducing more bootstrap utility classes.
Would you mind adding these styles in a scss file ?

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+  const { t: tCommon } = useTranslation(undefined, { keyPrefix: 'common' });
+
+  const cartImportType = useMemo(() => {
+    const inCart = cart.find((e) => e.id === trainScheduleSet.id);
```
The naming is confusing as we don't get a boolean from `find`
```suggestion
    const cartItem = cart.find((e) => e.id === trainScheduleSet.id);
```

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/types.ts:
```diff
@@ -0,0 +1,18 @@
+export type TrainScheduleSetImportType = 'copy' | 'reference';
+
+export type CartManagement = {
+  cart: Array<{ id: TrainScheduleSet['id']; type: TrainScheduleSetImportType }>;
```
I think this would be less confusing when using data with this type
```suggestion
  cart: Array<{ trainScheduleSetId: TrainScheduleSet['id']; importType: TrainScheduleSetImportType }>;
```
or 
```suggestion
  cart: Array<{ cartItemId: TrainScheduleSet['id']; importType: TrainScheduleSetImportType }>;
```

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+  const onClick = useCallback(() => {
+    if (disabled) return;
+    if (cartImportType) removeFromCart(trainScheduleSet.id);
+    else addToCart(trainScheduleSet.id, 'copy');
```
Just an open question as it is not in AC, do we want the new package in the cart to be in copy by default or in reference ? @maelysLeratRosso t.id, 'copy');

sim51 replied:
IMHO, it should be in reference. But the first icon is by copy ...

maelysLeratRosso replied:
I would say reference as well

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+    else addToCart(trainScheduleSet.id, 'copy');
+  }, [trainScheduleSet.id, cartImportType, addToCart, removeFromCart, disabled]);
+
+  const StatusIcon = useMemo(() => {
```
```suggestion
  const statusIcon = useMemo(() => {
```

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+        return <NoEntry variant="fill" className={className} />;
+      default:
+        return <PlusCircle className={className} />;
+    }
```
No really import but why use switch instead of an `if` ?


```suggestion
    if (cartImportType) return <NoEntry variant="fill" className={className} />;
    return <PlusCircle className={className} />;
```

sim51 replied:
My first version of this code had a definition on the reference branch.

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
Regarding the previous utility classes comment, we could remove them from all new files ?

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+            <span>{trainScheduleSet.name}</span>
+            {/* 
+              If the item is imported as a ref, we display the icon, 
+              as well if it is disable because it can only be a ref 
```
```suggestion
              as well if it is disabled because it can only be a ref 
```

sim51 replied:
Done, comment has been rewritten

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+            */}
+            {(cartImportType === 'reference' || disabled) && <Broadcast className="d-flex" />}
+          </div>
+          <div className="metadata">{trainScheduleSet.published ? 'published' : 'unpublished'}</div>
```
It seems like this part is not in the AC or mockup.
Maybe we could leave a todo here to replace it by the train_count

sim51 replied:
Done 
I checked with youness for the API and left a comment on the chan. For now we can only the total number of train schedule, we can't have the count of paced and missions (like in the design).

SharglutDev replied:
Thanks for the fix ! 

However, we can now see that, depending if the collapsable is open or not, the count text is not aligned : 
<img width="512" height="626" alt="image" src="https://github.com/user-attachments/assets/a9f1b583-5e4b-4dfd-9a9e-1e335aa41209" />


> [!NOTE]
> To summarize regarding this part for other reviewers : dates will maybe come later and for now we display to total of all types of trains (no difference between  paced train and unique trains)

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetsCart.tsx:
```diff
@@ -0,0 +1,55 @@
+  return (
+    <>
+      {trainScheduleSetsInCart.length === 0 && (
+        <p style={{ color: 'var(--grey50)', marginTop: '4rem' }}>{t('cartEmpty')}</p>
```
We try to avoid using inline style in the codebase

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+  addToCart,
+}: TrainScheduleSetCartItemProps) => {
+  const { t } = useTranslation();
+  const { t: tos } = useTranslation('operational-studies', { keyPrefix: 'importTrainscheduleSet' });
```
To harmonize with the naming you used in other components

```suggestion
  const { t: tCommon } = useTranslation();
  const { t } = useTranslation('operational-studies', { keyPrefix: 'importTrainscheduleSet' });
```

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+          getOptionValue={(set) => `${set}`}
+          getOptionIcon={(set) => (set === 'copy' ? <DesktopDownload /> : <Broadcast />)}
+          onChange={(set) => addToCart(trainScheduleSet.id, set)}
+          options={['reference', 'copy'] as TrainScheduleSetImportType[]}
```
This cast allow us to use any value for the options like 
`['ref', 'copy'] as TrainScheduleSetImportType[]`
and typescript won't complain, we could rather a a const above the component like this :
`const switchOptions: TrainScheduleSetImportType[] = ['reference', 'copy'];`

sim51 replied:
Done

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSet.tsx:
```diff
@@ -0,0 +1,129 @@
+  catalogEntry?: CatalogEntry;
+  importType: TrainScheduleSetImportType;
+};
+export const TrainScheduleSetCartItem = ({
```
```suggestion
};

export const TrainScheduleSetCartItem = ({
```

SharglutDev commented on front/src/modules/pathfinding/components/IncompatibleConstraints/IncompatibleConstrainstFilters.tsx:
Thank for the fix for this, maybe it could have been in a separate commit (all changes regarding `Collapsable`)

sim51 replied:
It's part of the commit [Fix! re-use the Collapse component](https://github.com/OpenRailAssociation/osrd/pull/14461/commits/59d8ecc67394f0186585771daba7bcbb6c58450f)  no?
And I needed to change it due to the refacto I did on the `Collapsable` component. 
Now the full line is in a button, so now in the `IncompatibleConstrainstFilters.tsx` we have a button within a button.

SharglutDev replied:
You're right my bad. You'll have to rename the commit though.

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+
+      aside.catalog {
+        background-color: var(--ambientB5);
+        border-right: 1px solid var(--color-black-25);
```
Was going to leave a comment to use the var from `_variables.scss` where we duplicated ui colors.
But maybe we should never have duplicated them at all ?
But in this file, we use both source of color var, we should stick to one

sim51 replied:
Hooo, indeed we have a problem here with color variables.

We have : 
 - sass variables 
 - sass that set css variables
 - css theme variables (with `--color-` pattern)
 
 And I see duplication of colors between css theme and sass variables. We need to have one source of truth.
 
In the meantime, I replaced the `var(--ambientYXX);` by `var(--color-ambientY-XX);` in this file. So now I use only one system.

SharglutDev replied:
Thanks for the harmonization. Added this topic to be discussed in the next front meeting

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+      aside.catalog {
+        background-color: var(--ambientB5);
+        border-right: 1px solid var(--color-black-25);
+        width: 34%;
```
Does this ratio have been checked with @thibautsailly ? 
Another way to handle the responsiveness could have been to shrink only the cart part which have a lot of empty space in large screen.

sim51 replied:
This ratio is a computation base on the width of each element in sketch (in fact it's somewhere between 33% & 34%) and it allows to have few responsive behavior. 

Without responsive design or guidelines, I made this choice.

If you shrink only the cart part, it works, but it means also grow it on big screen.
Sketch is based on a screen of 1440x970, and my screen is already bigger than that (1920x1080). 
That's why I prefer a ratio (and also my firebug opened on right side).

SharglutDev replied:
Just discussed with @thibautsailly, here what he would want : 
- the aside part width stay the same all the time
- the cart part shrinks until we can have 24px padding-inline around the cart
- after that the cart shrinks until ~350px
- after that we stop handling the responsive, osrd is not usable anymore

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+      padding: 1px 0;
+
+      aside.catalog {
+        background-color: var(--ambientB5);
```
```suggestion
        background-color: var(--ambientB10);
```

<img width="859" height="364" alt="Image" src="https://github.com/user-attachments/assets/e7c3472d-af50-4c96-8b85-e4c4902b1197" />

sim51 replied:
`var(--ambientB5)` is the right color to use here. 
You can pick the color on sketch and on the app, it's the same.

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+        .catalog-entry {
+          .collapse-header button {
+            font-weight: 600;
+            padding: 0.5rem;
```
We use rem only for `font-size`
```suggestion
            padding: 8px;
```

SharglutDev replied:
I can still see `rem` used in the file for "non font-size" styles

SharglutDev replied:
I think there is one `rem` remaining l.80 and l.98 but you removed a bit too much, the `font-size` should remain in `rem` :)

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+          }
+          .collapse-body {
+            background-color: var(--color-white-100);
+            padding-left: 1.625rem;
```
Because of this padding, we have a small zone without border at the bottom left of each collapsed element : 

<img width="233" height="187" alt="Image" src="https://github.com/user-attachments/assets/230eebef-6a0d-43cd-976a-62a43c3c1ed0" />

sim51 replied:
You means this 
<img width="560" height="356" alt="image" src="https://github.com/user-attachments/assets/c6c1ee88-7af1-4204-b693-500c4f231f96" />

SharglutDev replied:
Yes exactly. Discussed it with @thibautsailly, he harmonized the mockups which were not all the same regarding this zone.

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+            font-weight: 600;
+            padding: 0.5rem;
+
+            border-top: 1px solid var(--color-white-25);
```
I think this rather this value
```suggestion
            border-top: 1px solid var(--color-white-100);
```

<img width="771" height="170" alt="Image" src="https://github.com/user-attachments/assets/47fa72d9-3037-4335-b068-3347326fc39b" />

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+
+        .catalog-entry {
+          .collapse-header button {
+            font-weight: 600;
```
```suggestion
            font-weight: 600;
            color: var(--grey80);            
```

<img width="771" height="170" alt="Image" src="https://github.com/user-attachments/assets/31113e8e-62b5-44d9-a361-c38f192c769e" />

sim51 replied:
good catch

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+
+              .header {
+                padding: 0.5rem 1rem;
+                padding-left: 0;
```
Not a big change but this way we don't need to override the previous padding
```suggestion
                padding-block: 8px;
                padding-right: 16px;
```

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+                  line-height: 1.5rem;
+                  display: flex;
+                  align-items: center;
+                  gap: 0.25rem;
```
```suggestion
                  gap: 5px;
```

<img width="413" height="112" alt="Image" src="https://github.com/user-attachments/assets/8985ce70-3220-4458-9295-5890b405c54d" />

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+                  margin-right: 0.5rem;
+                }
+                .name {
+                  font-size: 1rem;
```
I think it's not needed ?
```suggestion
```

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+
+                  svg {
+                    display: block;
+                  }
```
Seems like this style is not working. Do you remember why we tried to add it ?
```suggestion
```

<img width="326" height="83" alt="Image" src="https://github.com/user-attachments/assets/7c916698-d0a3-43b6-8ba3-6442cb34484f" />

sim51 replied:
It was when I was fighting to vertical align icons from ui-icon, but agree, it's useless.

SharglutDev replied:
Yeah it's a pain this alignment :weary:

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+              .body {
+                padding: 0 1.5rem 1rem 2rem;
+                margin: 0;
+              }
+
+              &.disabled {
+                .header {
+                  color: var(--color-info-60);
+                }
```
```suggestion
                .header {
                  color: var(--color-info-60);
                  
                  .status {
                    color: var(--color-info-30);
                  }                  
                }
```

<img width="481" height="105" alt="Image" src="https://github.com/user-attachments/assets/0fa9a862-ef24-4089-8491-42f74873d182" />

sim51 replied:
Thanks! 
I know that I have a 8bit vision ...

SharglutDev replied:
Can't see the difference myself ! Just checked all colors to verify if they were the same :sweat_smile:

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+              color: var(--color-grey-50);
+            }
+          }
+          .remove {
```
Like often with ui icons, they are not vertically centered, we should a padding/margin to correct it

<img width="256" height="98" alt="Image" src="https://github.com/user-attachments/assets/c08076ab-a7d8-45ae-bd3e-a357a82330f0" />

sim51 replied:
A `display: flex` solved the issue

SharglutDev replied:
Oh nice ! Do you think it's a fix we could add in ui-icons template ?

If yes, we should leave a todo like we usually do so we can remember to remove this style when fixed in ui

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+            }
+          }
+          .remove {
+            padding-right: 1rem;
```
Because the padding of the card is 8px
```suggestion
            padding-right: 12px;
```

<img width="167" height="141" alt="Image" src="https://github.com/user-attachments/assets/6f0cdb39-2006-420a-8cdd-ef57ce75c4ef" />

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+              line-height: 1.5rem;
+            }
+            .catalog {
+              font-family: IBMPlexSans;
```
We don't need to add the font again
```suggestion
```
<img width="530" height="103" alt="Image" src="https://github.com/user-attachments/assets/d206eb95-b26e-44ad-99c8-61eba6edf0d8" />

sim51 replied:
Done

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+            .catalog {
+              font-family: IBMPlexSans;
+              font-size: 0.875rem;
+              font-style: semibold;
```
Not needed
```suggestion
```

sim51 replied:
Done

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+              font-weight: auto;
+              svg {
+                display: block !important;
+              }
```
Are these styles needed ? Tried to remove them and couldn't see the difference 🤔

sim51 replied:
In fact, all the `.type` and its children can be remove, it's part of the component in ui-core now.

sim51 replied:
Done

SharglutDev replied:
Could you remove the `type` class also in `TrainScheduleSetCartItem` l.29 ?

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,135 @@
+          gap: 1rem;
+          padding: 0.5rem;
+
+          .type {
```
Is it me or the border of the selected item is a bit too big compared to the mockup ?

App : 
<img width="110" height="87" alt="Image" src="https://github.com/user-attachments/assets/912a32d4-162a-476e-989d-8513656b2626" />

Mockup : 

<img width="126" height="112" alt="Image" src="https://github.com/user-attachments/assets/451e7c43-3dd8-479e-a870-3666304b9a02" />

sim51 replied:
The border is at 2px. If I set to 1px, here is the renderer:
<img width="502" height="163" alt="image" src="https://github.com/user-attachments/assets/37a83bec-1428-4d34-9c73-80b283ab5d82" />

It's hard to say on sketch, it uses box shadow to do it, and when I copy it the rule is not css valid ...
Whatever, this should be fixed on the `ui-core`. Can you open a ticket for it ?

SharglutDev replied:
Done : https://github.com/OpenRailAssociation/osrd/issues/14741

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetsCatalog.tsx:
```diff
@@ -0,0 +1,77 @@
+            <TriangleRight className="d-flex align-items-center justify-content-center mr-1" />
+          }
+        >
+          <span>{entry.name || 'empty'}</span>
```
I'm not sure this is possible for a catalog_entry name to be null here ?
We should rather assume that it won't with a ! or add a translated fallback.

What should we do @maelysLeratRosso ?

maelysLeratRosso replied:
A set should have a name yes
A package (local to a scenario) can have no set, but a set needs a name

sim51 replied:
Done

SharglutDev reviewed (commented):
Sorry for all the style comments. Went really with a 🔍 on the mockup

SharglutDev reviewed (commented):
Sorry for all the style comments. Went really with a 🔍 on the mockup

sim51 commented:
I just pushed a commit that fix the majority of your comments. 
My remaining task are :
- metadata on item in left menu (ie. published)
- change the useLoadCatalog to match the API plan (and retrieve the data
  for metadata)
- display the number of selected item in a catalog entry (forgot it)

sim51 commented:
Each comment has been solved

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetCartItem.tsx:
```diff
@@ -0,0 +1,54 @@
+      </div>
+      <div className="remove">
+        <button title={tCommon('delete')} onClick={() => removeFromCart(trainScheduleSet.id)}>
+          <NoEntry variant="fill" className="status flex-grow-0 d-flex align-items-center" />
```
Utility classes should in a css file (regarding this [comment](https://github.com/OpenRailAssociation/osrd/pull/14461#discussion_r2687165815))

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetMenuItem.tsx:
Same as above comment for all utility classes in this file

sim51 commented:
Found 2 styling issues :
- Menu item on the left column, doesn't take the full width 
- When a catalog entry is open, its background color changed

SharglutDev commented on front/public/locales/en/operational-studies.json:
```diff
@@ -54,6 +54,22 @@
+    "trainSchedulesCount_other": "{{count}} trains",
+    "trainScheduleSet": "Package",
+    "trainScheduleSetsSelected_one": "{{count}} selected version",
+    "trainScheduleSetsSelected_other": "{{count}} selected versions",
```
The mockup is stale for this text, it should say "package" instead of "version" ("paquet" for fr)

sim51 replied:
FR
```
"trainScheduleSetsSelected_one": "{{count}} paquet sélectionné",
 "trainScheduleSetsSelected_other": "{{count}} paquets sélectionnés",
 ```
 
 EN
 ```
"trainScheduleSetsSelected_one": "{{count}} package selected",
"trainScheduleSetsSelected_other": "{{count}} packages selected",
 ```

SharglutDev commented on front/src/styles/scss/applications/operationalStudies/_importTrainScheduleSets.scss:
```diff
@@ -0,0 +1,144 @@
+              font-weight: 600;
+              padding-block: 8px;
+              padding-left: 8px;
+              padding-right: 16px;
```
nit: You can probably merge them all in a `padding`

SharglutDev commented:
>     * When a catalog entry is open, its background color changed

Good catch, indeed it was one of the inconsistency in the mockups we discussed yesterday with @thibautsailly and he fixed it

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetsCart.tsx:
Overflow doesn't work correctly when there are a lot of items in the cart

<img width="680" height="888" alt="Image" src="https://github.com/user-attachments/assets/83cd6a2f-75a7-46cc-b5b2-caf000136e3c" />

sim51 replied:
Done, it was a missing `overflow:scroll` on the cart panel.
Moreover I added a padding block because the last element was touching the bottom border of the panel.

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetCatalogDialog.tsx:
```diff
@@ -0,0 +1,124 @@
+    <Dialog
+      fullscreen
+      className={cx('import-trainscheduleset', error && 'with-error')}
+      header={<h5>{t('title')}</h5>}
```
We need to remove the `h5` bottom margin

<img width="406" height="113" alt="Image" src="https://github.com/user-attachments/assets/4b231080-181f-49d4-99d2-5efce984c41b" />

sim51 replied:
I created the issue #14745 because it should be integrated in ui-core

SharglutDev reviewed (commented):
I think we will be good to go after these last comments !

SharglutDev commented on front/src/applications/operationalStudies/views/Scenario/components/ImportTrainScheduleSets/TrainScheduleSetMenuItem.tsx:
```diff
@@ -49,22 +49,22 @@ const TrainScheduleSetMenuItem = ({
   return (
     <button
-      className={cx('trainscheduleset-card d-flex flex-column', disabled && 'disabled')}
+      className={cx('trainscheduleset-card', disabled && 'disabled')}
```
nit: this should also work
```suggestion
      className={cx('trainscheduleset-card', { disabled })}
```

SharglutDev reviewed (approved):
Lgtm and tested, great job ! ✔️
